src/com/MPISinglePortsCommunication.cpp

```diff
@@ -0,0 +1,207 @@
+  DEBUG("Accept connection at " << _portName);
+
+  // Connect the first peer, s.t. we can exchange some information about the other side
+  MPI_Comm communicator;
+  MPI_Comm_accept(const_cast<char *>(_portName.c_str()), MPI_INFO_NULL, 0, MPI_COMM_SELF, &communicator);
+  DEBUG("Accepted connection at " << _portName);
+
+  int    requesterRank      = -1;
+  size_t requesterCommunicatorSize = 0;
+
+  // Exchange information to which rank I am connected and which communicator size on the other side
+  MPI_Recv(&requesterRank,             1, MPI_INT, 0, 42, communicator, MPI_STATUS_IGNORE);
+  MPI_Recv(&requesterCommunicatorSize, 1, MPI_INT, 0, 42, communicator, MPI_STATUS_IGNORE);
+  MPI_Send(&acceptorRank,              1, MPI_INT, 0, 42, communicator);
+  
+  CHECK(requesterCommunicatorSize > 0, "Requester communicator size has to be > 0!");
+  _communicators[requesterRank] = communicator;
+  
+  // Connect all other peers
+  for (size_t i = 1; i < requesterCommunicatorSize; ++i) {
+    MPI_Comm_accept(const_cast<char *>(_portName.c_str()), MPI_INFO_NULL, 0, MPI_COMM_SELF, &communicator);
+    DEBUG("Accepted connection at " << _portName);
+
+    MPI_Recv(&requesterRank,             1, MPI_INT, 0, 42, communicator, MPI_STATUS_IGNORE);
+    MPI_Recv(&requesterCommunicatorSize, 1, MPI_INT, 0, 42, communicator, MPI_STATUS_IGNORE);
+    MPI_Send(&acceptorRank,              1, MPI_INT, 0, 42, communicator);
+
+    CHECK(requesterCommunicatorSize == _communicators.size(),
+          "Requester communicator sizes are inconsistent!");
+    CHECK(_communicators.count(requesterRank) == 0,
+          "Duplicate request to connect by same rank (" << requesterRank << ")!");
+
+    _communicators[requesterRank] = communicator;
+  }
+  
+  _isConnected = true;
+}
+
+/// requesterCommunicatorSize is not used, since connection is always made on the entire communicator
+void MPISinglePortsCommunication::acceptConnectionAsServer(
+    std::string const &acceptorName,
+    std::string const &requesterName,
+    int                acceptorRank,
+    int                requesterCommunicatorSize)
+{
+  TRACE(acceptorName, requesterName, acceptorRank, requesterCommunicatorSize);
+  assertion(not isConnected());
+
+  _isAcceptor = true;
+
+  const std::string addressFileName("." + requesterName + "-" + acceptorName + ".address");
+  Publisher::ScopedChangePrefixDirectory scpd(_addressDirectory);
+  ScopedPublisher                        p(addressFileName);
+  if (utils::MasterSlave::_rank == 0) { // only master opens a port
+    MPI_Open_port(MPI_INFO_NULL, const_cast<char *>(_portName.data()));
+    p.write(_portName);
+    DEBUG("Accept connection at " << _portName);
+  }
+  
+  MPI_Comm communicator;
+  MPI_Comm_accept(const_cast<char *>(_portName.c_str()), MPI_INFO_NULL, 0,
+                  utils::Parallel::getGlobalCommunicator(), &communicator);
+  DEBUG("Accepted connection at " << _portName);
+  _communicators[0] = communicator; // all comms are the same
```

Comment:
_Optional:_ you could assert this in debug builds using:
```cpp
std::all(_communicators.cbegin(), _communicator.cend(), [communicator](const auto& kv){
return kv.second == communicator;
})
```

Reply:
`_communicator.size() == 1` here always. The comment rather says, that all ranks are addressed using the same communicator (when using the `acceptAsServer` and `requestAsClient` methods.